Factor UnboundedQuantityValue out of QuantityValue

brightbyte opened the pull request:
Code experiment exploring the possibility to use subclassing to model unbounded vs bounded quantities.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	 * @throws IllegalValueException
+	 */
+	public static function newFromNumber( $number, $unit = '1', $upperBound = null, $lowerBound = null ) {
+		if ( $number instanceof QuantityValue ) {
```
This may already have a unit and bounds, which will be ignored. I don't think this is helpful.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	 * @return self
+	 */
+	public static function newFromDecimal( $number, $unit = '1', $upperBound = null, $lowerBound = null ) {
+		if ( $number instanceof QuantityValue ) {
```
As above.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	 * @return string
+	 */
+	public static function getType() {
+		return 'quantity'; // FIXME: should this have a different type id? wehat will that break?
```
I think: both is the same, the fact that we use subclasses has nothing to do with the data model. It's to make the code nicer.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	public function getArrayValue() {
+		return array(
+			'amount' => $this->amount->getArrayValue(),
+			'unit' => $this->unit,
```
Could call the super class here.

brightbyte replied:
I prefer not to, to keep this symmetrical and consistent with newFromArray

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -5,9 +5,11 @@
+ * Class representing a quantity with associated unit.
  * The amount is stored as a @see DecimalValue object.
  *
+ * For quantities with a known uncertainty interval, see BoundedQuantityValue.
```
Have an actual `@see`?

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -93,36 +67,32 @@ public function __construct( DecimalValue $amount, $unit, DecimalValue $upperBou
 	/**
 	 * @see newFromNumber
 	 *
 	 * @deprecated since 0.1, use newFromNumber instead
```
Breaking change anyway, can easily be removed now.

brightbyte replied:
what can be removed?

brightbyte replied:
oh, the deprecated method, right.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+		// note: we really want floor( log10( $precision ) ), but have to account for
+		// small errors made in the floating point operations above.
+		// @todo: use bcmath (via DecimalMath) to avoid this if possible
+		$orderOfUncertainty = floor( log10( $precision + 0.0000000005 ) );
```
Let's please not re-introduce this odd code. Let's rewrite this where it is needed.

brightbyte replied:
Let's please not rewrite this odd code, it was rather tricky to get right; we can factor it out into a separate class, but let's do that in a separate change.

Let's also keep in mind that the transform() method is central to unit conversion. Even if it's not really used at the moment, we should keep it around.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	 *
+	 * @return int
+	 */
+	public function getSignificantFigures() {
```
Not used anywhere, can simply be removed.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	 * @throws InvalidArgumentException
+	 * @return self
+	 */
+	public function transform( $newUnit, $transformation ) {
```
Not used anywhere. Does not belong to a value object. Please remove.

brightbyte replied:
We will need it for unit conversion, for formatting and indexing. Deleting tricky code is asking for trouble. I'm open to moving it though, see above.

thiemowmde commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,381 @@
+	public static function newFromArray( $data ) {
+		self::requireArrayFields( $data, array( 'amount', 'unit', 'upperBound', 'lowerBound' ) );
+
+		return new static(
```
This could call the parent class when the two fields are missing (for convenience).

brightbyte commented on src/DataValues/QuantityValue.php:
```diff
@@ -237,128 +202,6 @@ public function getAmount() {
-	 *
-	 * @return int
-	 */
-	public function getSignificantFigures() {
```
I'm happy to discuss factoring the significant digits stuff out, but let's please do that in a later pull request, so we keep steps small and don't bock one change because of discussion about the other.
I agree that it should target the same major release, though.

JeroenDeDauw commented on composer.json:
```diff
@@ -25,6 +25,10 @@
 	},
+	"require-dev": {
+		"mediawiki/mediawiki-codesniffer": "0.7.2",
+		"phpunit/phpunit": "4.8.24"
```
Why disallow newer versions?

brightbyte replied:
This is just copy&paste from core. I don't care much.

JeroenDeDauw replied:
Copy pasting from MediaWiki does not strike me as a good idea :)

JeroenDeDauw commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,352 @@
+ * The fact that we use subclassing to model the bounded vs the unbounded case should be
+ * considered an implementation detail.
+ *
+ * @since 0.1
```
?

brightbyte replied:
Omnomnom, copy pasta!

JeroenDeDauw commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,352 @@
+	/**
+	 * Constructs a new QuantityValue object, representing the given value.
+	 *
+	 * @since 0.1
```
?

JeroenDeDauw commented on src/DataValues/BoundedQuantityValue.php:
```diff
@@ -0,0 +1,352 @@
+	private $lowerBound;
+
+	/**
+	 * Constructs a new QuantityValue object, representing the given value.
```
![42be4aef-36a8-a182-1851172](https://cloud.githubusercontent.com/assets/146040/16357355/a56954a8-3af4-11e6-9e6e-aaf3ed0dadc0.jpg)

JeroenDeDauw commented on composer.json:
```diff
@@ -25,6 +25,9 @@
 		"data-values/common": "~0.3.0|~0.2.0"
 	},
+	"require-dev": {
+		"phpunit/phpunit": "~4.8"
```
Why not the current major version of PHPUnit?

brightbyte replied:
I prefer for all our libraries to use the same version of tools and libraries. Otherwise, it becomes confusing what feature you can use where. If we want to move to 5.x, let's do that for all our stuff.

thiemowmde replied:
Should be a separate patch, please.

brightbyte replied:
I added it to this patch so I can uses phpunit to test this patch :)

JeroenDeDauw replied:
Clearly it was possible to test the code before you made this change?

brightbyte replied:
Yes, if you happened to have the correct version of phpunit installed. If you didn't, you had to guess which version would work.

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -8,28 +8,19 @@
+ * @see UnboundedQuantityValue For quantities with no known uncertainty interval.
  * For simple numeric amounts use @see NumberValue.
  *
- * @since 0.1
```
Did the fact that the class exists since 0.1 changed?

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -91,155 +70,59 @@ public function __construct( DecimalValue $amount, $unit, DecimalValue $upperBou
 	 * to the rules defined by @see DecimalValue.
 	 *
-	 * @since 0.1
+	 * @since 0.7
```
As with all `@since` tags in this file, this should not change because the method signature did not changed, as far as I can see. That was the whole point of renaming the two classes, wasn't it?

brightbyte replied:
THis is a sideeffect of renaming this back-and-forth. In the first version of this change, this code was in the new BoundedQuantityValue class. I'll fix all the since tags.

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -91,155 +70,59 @@ public function __construct( DecimalValue $amount, $unit, DecimalValue $upperBou
+	 * @since 0.7
 	 *
-	 * @param string|int|float|DecimalValue $amount
+	 * @param string|int|float|DecimalValue|UnboundedQuantityValue $number
```
I believe this is a mistake. We talked about this and decided to not allow using an other QuantityValue as a number because you can pass unit and bounds twice that way.

Same a few lines below.

brightbyte replied:
ok

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -91,155 +70,59 @@ public function __construct( DecimalValue $amount, $unit, DecimalValue $upperBou
 	 * @param string|int|float|DecimalValue|null $lowerBound
 	 *
-	 * @return self
+	 * @return QuantityValue
```
A mistake that happened when renaming the files? Should return `self`.

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -91,155 +70,59 @@ public function __construct( DecimalValue $amount, $unit, DecimalValue $upperBou
 		list( $amount, $unit, $upperBound, $lowerBound ) = unserialize( $data );
+		$amount = DecimalValue::newFromArray( $amount );
+		$upperBound = DecimalValue::newFromArray( $upperBound );
+		$lowerBound = DecimalValue::newFromArray( $lowerBound );
```
I wonder what happened here? Did the serialization format changed?

brightbyte replied:
Uh, I removed this, why is it back?? Damn rebase hell :(

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -473,14 +322,19 @@ public function getArrayValue() {
 	 * @throws IllegalValueException
 	 */
 	public static function newFromArray( $data ) {
+		if ( !isset( $data['upperBound'] ) && isset( $data['lowerBound'] ) ) {
```
There is an `!` missing. A test should catch all possible cases.

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,334 @@
+ * The fact that we use subclassing to model the bounded vs the unbounded case should be
+ * considered an implementation detail.
+ *
+ * @since 0.1
```
0.7

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,334 @@
+	/**
+	 * @see newFromNumber
+	 *
+	 * @deprecated since 0.1, use newFromNumber instead
```
Introducing a deprecated method into a new class is odd.

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,334 @@
+	 */
+	public function unserialize( $data ) {
+		list( $amount, $unit ) = unserialize( $data );
+		$amount = DecimalValue::newFromArray( $amount );
```
Again, I wonder why you introduced this. Probably for type safety and early exceptions, but the `__constructor` and `asDecimalValue` already do this.

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,334 @@
+	/**
+	 * Returns the quantity object.
+	 *
+*@see DataValue::getValue
```
Indention.

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,334 @@
+	 * @return string
+	 */
+	public function serialize() {
+		return serialize( array_values( $this->getArrayValue() ) );
```
I just realized that this **is** a change to the serialization format. Side effects?

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -2,9 +2,10 @@
 
 namespace ValueFormatters;
 
+use DataValues\QuantityValue;
```
Odd non-alphabetic reordering.

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -26,6 +27,24 @@ class QuantityFormatter extends ValueFormatterBase {
+	 * Value for the OPT_SHOW_UNCERTAINTY_MARGIN indicating that the uncertainty margin
+	 * should be shown if we are displaying a non-exact QuantityValue.
+	 */
+	const SHOW_UNCERTAINTY_MARGIN_IF_NOT_ZERO = 'if not zero';
```
1. I believe this must be a separate patch.
2. Using `…&showQuantityUncertaintyMargin=if%20not%20zero` in an URL is odd. Please avoid spaces.

brightbyte replied:
as discussed

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -150,43 +172,50 @@ protected function formatQuantityValue( QuantityValue $quantity ) {
+	private function getRoundingExponent( UnboundedQuantityValue $quantity, $roundingMode ) {
+		if ( is_int( $roundingMode ) ) {
+			// round to the given exponent
+			return $roundingMode;
```
Support for integers in a boolean option that's called "apply rounding"? Sorry, not convinced. Especially because the numbers are base-10 exponents, and conflict with what lower and upper bound are meant to do.

Again, this looks like a new feature that **must** be in it's own pull request.

brightbyte replied:
This is not new, we have been supporting int values for this option for years.

thiemowmde replied:
The code I'm complaining about is green, marked as new. I wonder what made it necessary to rearrange it. Did you re-introduced the formerly removed boolean options?

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -197,16 +226,22 @@ private function getRoundingExponent( QuantityValue $quantity ) {
+		}
+
+		// TODO: never round to 0! See bug #56892
+		$roundedMargin = $this->decimalMath->roundToExponent( $margin, $roundingExponent );
```
Applying the same (guessed) rounding to both the base number and the margin never was correct.

brightbyte replied:
Giving a margin of uncertainty with more precision than the value itself is also never correct.
But I agree that we should just show this as entered, correct or not.

thiemowmde commented on tests/DataValues/QuantityValueTest.php:
```diff
@@ -290,17 +261,17 @@ public function transformProvider() {
 		};
 
 		return array(
-			 0 => array( QuantityValue::newFromNumber( '+10',   '1', '+11',  '+9' ),   $identity, QuantityValue::newFromNumber(   '+10',    '?',   '+11',    '+9' ) ),
```
These whitespace changes make it hard to review this.

thiemowmde commented:
In general, I would love to start over with a lot of the implementation details in here, but can't as long as we are stuck with all this and not allowed to remove even //unused// code before we continue. As long as there is any code in our code base that turns 1.5±1 into 2±1 and 2.9±1.9 into 3±2 we are doing it **wrong**.

brightbyte commented:
Updated the patch as discussed. A few notes:

* Removed unused getSignificantDigits method. Note that transform() will be used by https://gerrit.wikimedia.org/r/#/c/296962/
* Postponed the changing rounding behavior according to T95425 resp T105623 until we have QuantityMath
* New options in the formatter stay in, since we need new behavior for unbound quantities anyway.

brightbyte commented:
*sigh* after the rebase, lots of tests fail. no time to look into that now.

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -473,14 +342,22 @@ public function getArrayValue() {
 	 * @param mixed $data
 	 *
-	 * @return self
+	 * @return UnboundedQuantityValue
```
Callers should be aware that this can return `UnboundedQuantityValue|self`.

brightbyte replied:
Callers should only care that the result is guaranteed to be an UnboundedQuantityValue, and not guaranteed to be a QuantityValue.

thiemowmde replied:
The formatter **must** know this can return a bound value, otherwise we can not display the uncertainty any more.

brightbyte replied:
It does know that. Anything that can return an UnboundedQuantityValue can also return a QuantityValue.

thiemowmde replied:
The documentation should say that then.

thiemowmde commented on src/DataValues/QuantityValue.php:
```diff
@@ -473,14 +342,22 @@ public function getArrayValue() {
 	 * @throws IllegalValueException
 	 */
 	public static function newFromArray( $data ) {
+		if ( !isset( $data['upperBound'] ) || !isset( $data['lowerBound'] ) ) {
```
Must be `&&`. As far as I remember I said this before: Please add tests for at least one possible combination to cover this.

brightbyte replied:
With &&, if only one of the fields is missing, we'll die with a fatal error. Is that desirable?

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,323 @@
+ * The fact that we use subclassing to model the bounded vs the unbounded case should be
+ * considered an implementation detail.
+ *
+ * @since 0.7
```
0.7 is already released.

thiemowmde commented:
Would be easier to have the parser changes in a separate patch. I probably have not said that clear enough when we discussed this.

thiemowmde commented:
+706 −294. :-( Can we please, please split this patch?

1. Add the new base class. Done.
2. Change the formatter to accept the base class.
3. Change the parser to return the base class.

thiemowmde commented:
To what does "Removed unused getSignificantDigits method" refer to? There is no such method.

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -26,6 +27,24 @@ class QuantityFormatter extends ValueFormatterBase {
+	 * Value for the OPT_SHOW_UNCERTAINTY_MARGIN indicating that the uncertainty margin
+	 * should be shown if we are displaying a non-exact QuantityValue.
+	 */
+	const SHOW_UNCERTAINTY_MARGIN_IF_NOT_ZERO = 'if-not-zero';
```
This means we will mix camel case with dash style: `showQuantityUncertaintyMargin=if-not-zero`. I wonder if we should have a style guide for this.

brightbyte replied:
I kind of like camel case for names and dashes for values... maybe I just have a strange taste.

thiemowmde commented on src/ValueFormatters/QuantityFormatter.php:
```diff
@@ -178,23 +219,25 @@ protected function formatNumber( QuantityValue $quantity ) {
+			// round to the order of uncertainty (QuantityValue only)
+			return $quantity->getOrderOfUncertainty();
 		} else {
-			return (int)$this->options->getOption( self::OPT_APPLY_ROUNDING );
```
This looks like an unrelated breaking change. Before the code accepted numeric strings. Now it seems it ignores the option if e.g. `"2"` is provided as a string.

thiemowmde commented on src/ValueParsers/QuantityParser.php:
```diff
@@ -2,10 +2,11 @@
 
 namespace ValueParsers;
 
+use DataValues\QuantityValue;
```
Why did this moved?

thiemowmde commented on src/DataValues/UnboundedQuantityValue.php:
```diff
@@ -0,0 +1,323 @@
+			$number = $default;
+		}
+
+		if ( $number instanceof UnboundedQuantityValue ) {
```
We decided this should not be done.
